### PR TITLE
fully validate multiaddr byte representations on construction

### PR DESCRIPTION
**Pull request by Stebalien**
Given that we have to modify transcoders anyways, this patch also moves some
repetitive logic out of these transcoders into this library.

### CODE REVIEW

File: codec.go [hunk 1]
```diff
@@ -50,6 +50,9 @@ func stringToBytes(s string) ([]byte, error) {
 		if err != nil {
 			return nil, fmt.Errorf("failed to parse %s: %s %s", p.Name, sp[0], err)
 		}
+		if p.Size < 0 { // varint size.
+			b.Write(CodeToVarint(len(a)))
```
Comment (whyrusleeping) [inline codec.go]: maybe explicitly throw away the error values?

File: transcoders.go [hunk 2]
```diff
@@ -121,61 +131,36 @@ func onionBtS(b []byte) (string, error) {
...
-	if len(b) != size {
-		return "", errors.New("inconsistent lengths")
-	}
-	if size == 0 {
-		return "", errors.New("invalid length")
-	}
-	s := string(b)
-	s = s[1:] // remove starting slash
-	return s, nil
+	return string(b), nil
```
Comment (whyrusleeping) [inline transcoders.go]: reasoning here?

File: transcoders.go [hunk 3]
```diff
@@ -121,61 +131,36 @@ func onionBtS(b []byte) (string, error) {
...
 
-var TranscoderUnix = NewTranscoderFromFunctions(unixStB, unixBtS)
+var TranscoderUnix = NewTranscoderFromFunctions(unixStB, unixBtS, nil)
 
 func unixStB(s string) ([]byte, error) {
-	// the address is the whole remaining string, prefixed by a varint len
-	size := CodeToVarint(len(s))
-	b := append(size, []byte(s)...)
-	return b, nil
+	return []byte(s), nil
```
Comment (whyrusleeping) [inline transcoders.go]: are we dropping the varints for some reason? I assume youre not changing the binary format, so where is this being moved to?

File: transcoders.go (hunk 2 above)
Comment (Stebalien) [reply transcoders.go]: I wanted it to be symmetric (that is, the bytes to string function should be an inverse of the string to bytes function). I'm now handling this in `bytesToString` (as we do in `stringToBytes`).

File: transcoders.go (hunk 3 above)
Comment (Stebalien) [reply transcoders.go]: We're now doing this in stringToBytes and bytesToString. Before, we were doing this for *every* protocol with a variable length size (and we were doing the exact same thing. On the other hand, we had *other* code that assumed that we were using varint length prefixes so I figured we might as well just centralize this logic and simplify things.

This change and the path one are both to pave the way for nice multiaddr constructors/decoders (so we can stop parsing strings).

File: transcoders.go (hunk 2 above)
Comment (Stebalien) [reply transcoders.go]: We could also not prepend a slash here:

https://github.com/multiformats/go-multiaddr/blob/84fb105635d33c95b41a9385064329b663f6f56f/codec.go#L43

**Pull request merged**